modules/Bio/EnsEMBL/DBSQL/BiotypeAdaptor.pm

```diff
@@ -0,0 +1,190 @@
+use Bio::EnsEMBL::Biotype;
+
+use strict;
+use warnings;
+
+use parent qw(Bio::EnsEMBL::DBSQL::BaseAdaptor);
+
+=head2 _tables
+
+  Arg [1]    : none
+  Description: PROTECTED implementation of superclass abstract method.
+               Returns the names, aliases of the tables to use for queries.
+  Returntype : list of listrefs of strings
+  Exceptions : none
+
+=cut
+
+sub _tables {
+  my $self = shift;
+
+  return (['biotype', 'b']);
+}
+
+=head2 _columns
+
+  Arg [1]    : none
+  Example    : none
+  Description: PROTECTED implementation of superclass abstract method.
+               Returns a list of columns to use for queries.
+  Returntype : list of strings
+  Exceptions : none
+
+=cut
+
+sub _columns {
+  my $self = shift;
+
+  return ('b.biotype_id', 'b.name', 'b.object_type', 'b.db_type', 'b.attrib_type_id', 'b.description', 'b.biotype_group', 'b.so_acc');
+}
+
+=head2 _objs_from_sth
+
+  Arg [1]    : StatementHandle $sth
+  Example    : none
+  Description: PROTECTED implementation of abstract superclass method.
+               responsible for the creation of ProteinFeatures
+  Returntype : listref of Bio::EnsEMBL::Biotype objects
+  Exceptions : none
+
+=cut
+
+sub _objs_from_sth {
+  my ($self, $sth) = @_;
+
+  my ($dbID, $name, $object_type, $db_type, $attrib_type_id, $description, $biotype_group, $so_acc);
+
+  $sth->bind_columns(\$dbID, \$name, \$object_type, \$db_type, \$attrib_type_id, \$description, \$biotype_group, \$so_acc);
+
+  my @biotypes;
+
+  while($sth->fetch()) {
+    push( @biotypes,
+      my $feat = Bio::EnsEMBL::Biotype->new_fast( {
+         'dbID'           => $dbID,
+         'name'           => $name,
+         'object_type'    => $object_type,
+         'db_type'        => $db_type,
+         'attrib_type_id' => $attrib_type_id,
+         'description'    => $description,
+         'biotype_group'  => $biotype_group,
+         'so_acc'         => $so_acc,
+      } )
+    );
+  }
+
+  return \@biotypes;
+}
+
+
+=head2 fetch_by_name_object_type
+
+  Arg [1]    : String $name
+               The name of the biotype to retrieve
+  Arg [2]    : String $object_type
+               The object type of the biotype to retrieve (gene or transcript)
+  Example    : $biotype = $biotype_adaptor->fetch_by_name_object_type('gene', 'protein_coding');
+  Description: Retrieves a biotype object from the database via its combined key (name, object_type).
+               If the Biotype requested does not exist in the database, a new Biotype object is
+               created with the provided name and object_type to be returned.
+  Returntype : Bio::EnsEMBL::Biotype
+  Exceptions : none
+
+=cut
+
+sub fetch_by_name_object_type {
+  my ($self, $name, $object_type) = @_;
+
+  my $constraint = "b.name = ? AND b.object_type = ?";
+  $self->bind_param_generic_fetch($name, SQL_VARCHAR);
+  $self->bind_param_generic_fetch($object_type, SQL_VARCHAR);
+  my ($biotype) = @{$self->generic_fetch($constraint)};
+
+  if (!defined $biotype) {
```

Comment:
I think this might be too helpful. It can lead to working with biotypes that aren't in the DB without knowing it. I would prefer an exception and a more wary calling code.

Reply:
So it's looks this is required which means I can't raise an exception...
Do you think a warning would be a good compromise?

Reply:
I suspect warnings would upset genebuild. Document the need instead?